docs/developer/best-practices/navigation.asciidoc

```diff
@@ -0,0 +1,201 @@
+[[kibana-navigation]]
+== Routing, Navigation and URL
+
+{kib} platform provides a set of tools to help developers built consistent experience around routing and browser navigation.
+Some of that tooling is inside the `core`, some is available as part of various plugins.  
+
+The purpose of this guide is to give you (a {kib} contributor}) a high-level overview of available tools and to explain common approaches for handling routing,
+and browser navigation.
+
+This guide covers following topics:
+
+* <<navigating-between-kibana-apps>>
+* <<deep-linking>>
+* <<routing>>
+* <<history-and-location>>
+* <<state-sync>>
+* <<preserve-state>>
+
+
+[[navigating-between-kibana-apps]]
+=== Navigating between {kib} apps
+
+{kib} is a single page application and there is a set of simple rules developers should follow
+to make sure there is no page reload when navigating from one place in {kib} to another. 
+
+**For example**, navigation using native browser apis:  
+
+[source,js]
+----
+window.location.href = `/dashboard/my-dashboard`; // (try to avoid this)
+----
+
+would cause a full page reload.
+
+To navigate between different {kib} apps without a page reload there are apis in `core`:
+
+* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetoapp.md[core.application.navigateToApp]
+* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetourl.md[core.application.navigateToUrl]
+
+*Rendering a link to a different {kib} app on its own would also cause a full page reload:*
+
+[source,typescript jsx]
+----
+const myLink = () => <a href={`/dashboard/my-dashboard`}> Go to Dashboard </a>; 
+----
+
+A workaround would be to handle a click, prevent browser navigation and use `core.application.navigateToUrl` api:
+
+[source,typescript jsx]
+----
+const MySPALink = () => 
+    <a 
+        href={`/dashboard/my-dashboard`} 
+        onClick={(e) => {
+            e.preventDefault();
+            core.application.navigateToUrl(`/dashboard/my-dashboard`); 
+        }}> 
+        Go to Dashboard 
+    </a>; 
+----
+
+
+It is too much boilerplate to do this for each {kib} link in your app, so there is a handy HOC that helps with it: 
+{kib-repo}tree/{branch}/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49[RedirectAppLinks].
+
+[source,typescript jsx]
+----
+const MyApp = () => 
+  <RedirectAppLinks application={core.application}>
+    {/*...*/}
+        <a href={`/dashboard/my-dashboard`}>Go to Dashboard</a> {/* navigations using this link will happen in SPA friendly way */}
```

Comment:
Links inside the `RedirectAppLinks` component also need the basePath:

https://github.com/elastic/kibana/blob/b7057b7b22ae223483249f108eeee295fa1516ab/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L37-L42